airflow/providers_manager.py

```diff
@@ -46,42 +50,136 @@ class ProvidersManager:
+        in case you have both sources and packages installed, the providers will be loaded from
+        the "airflow" sources rather than from the packages.
+        """
+        import airflow.providers
+
+        try:
+            for path in airflow.providers.__path__:
+                self.__add_provider_info_from_local_source_files_on_path(path)
+        except Exception as e:  # noqa pylint: disable=broad-except
+            log.warning("Error when loading 'provider.yaml' files from airflow sources: %s", e)
+
+    def __add_provider_info_from_local_source_files_on_path(self, path) -> None:
+        """
+        Finds all the provider.yaml files in the directory specified.
+        :param path: path where to look for provider.yaml files
+        """
+        root_path = path
+        for folder, _, files in os.walk(path):
+            for filename in fnmatch.filter(files, "provider.yaml"):
+                self.__add_provider_info_from_local_source_file(filename, folder, root_path)
+
+    def __add_provider_info_from_local_source_file(self, filename, folder, root_path) -> None:
+        """
+        Parses found provider.yaml file and adds found provider to the list of all providers.
+        :param filename: name of the provider.yaml file
+        :param folder: folder where the provider.yaml file is
+        :param root_path: root path where we started to search ("airflow/providers" folder from the
+                          local sources
+        """
+        try:
+            with open(os.path.join(folder, filename)) as provider_yaml_file:
+                provider_info = yaml.safe_load(provider_yaml_file.read())
+            self._validator.validate(provider_info)
+            package_name = "apache-airflow-providers" + folder[len(root_path) :].replace(os.sep, "-")
+            version = provider_info['versions'][0]
+            if package_name not in self._provider_set:
+                self._provider_set.add(package_name)
+                self._provider_list.append((package_name, version, provider_info))
+            else:
+                log.warning(
+                    "The providers for package '%s' could not be registered because providers for that "
+                    "package name have already been registered",
+                    package_name,
+                )
+        except Exception as e:  # noqa pylint: disable=broad-except
+            log.warning("Error when loading '%s/%s': %s", folder, filename, e)
 
     @property
     def providers(self):
         """Returns information about available providers."""
         return self._provider_directory
+
+
+# After we move to python 3.9 + we can get rid of this and use @cache from the functools
+cache = lru_cache(maxsize=None)
+
+
+@cache
+def get_providers_manager() -> ProvidersManager:  # noqa
+    """Returns singleton instance of ProvidersManager"""
+    return ProvidersManager()
```

Comment:
We could do this more "formally" as Daniel did in https://github.com/apache/airflow/commit/3ca11eb9b02a2c2591292fd6b76e0e98b8f22656#diff-681de8974a439f70dfa41705f5c1681ecce615fac6c4c715c1978d28d8f0da84R63

Reply:
Surely, There is no standard way of getting singleton in Python but if that's the pattern that is already used in Airflow, no problem with that.